### PR TITLE
Added Pointer Networks implementation based on the Vocabulary Approach

### PR DESCRIPTION
**Pull request by Zarana-Parekh**
Added the pointer networks implementation for summarization models. This is based on the vocabulary approach (for copying probability) as described in:
See, Abigail, Peter J. Liu, and Christopher D. Manning. "Get to the point: Summarization with pointer-generator networks." arXiv preprint arXiv:1704.04368 (2017).
It also includes adapting the coverage mechanism of Sockeye to work with the pointer networks model as given in the paper.

#### Pull Request Checklist ##
- [] Changes are complete (if posting work-in-progress code, prefix your pull request title with '[WIP]'
until you can check this box.
- [x] Unit tests pass (`pytest`)
- [ ] Were system tests modified? If so did you run these at least 5 times to account for the variation across runs?
- [ ] System tests pass (`pytest test/system`)
- [ ] Passed code style checking (`./style-check.sh`)
- [x] You have considered writing a test
- [x] Updated major/minor version in `sockeye/__init__.py`. Major version bump if this is a backwards incompatible change.
- [x] Updated CHANGELOG.md

By submitting this pull request, I confirm that my contribution is made as an Amazon employee.

### CODE REVIEW

**Comment by Zarana-Parekh** [comment]
ping @mjpost @fhieber @tdomhan

**Comment by mjpost** [comment]
Hi Zarana—thanks, will look at this soon.

File: sockeye/training.py [hunk 1]
```diff
@@ -160,16 +168,68 @@ def sym_gen(seq_lens):
...
+                    ext_vocab_size = self.config.config_embed_source.vocab_size + self.config.max_oov_words
+
+                    # prob_vocab: (batch_size * target_seq_len, vocab_size)
+                    # prob_source: (batch_size * target_seq_len, source_len)
+                    prob_vocab, prob_source = self.output_layer(target_decoded, context=context,
+                                                                          attention=attention,
+                                                                          target_embed=target_embed)
+
+                    # to correctly add source word probabilities based on the source word indices in the vocabulary
+                    # generate corresponding indices matrix
```
Comment (mjpost) [inline sockeye/training.py]: I understand the net effect of the code below, but following the details is difficult. Can you add a comment above each line, describing what exactly that line is doing, and why? Basically, a detailed walk-through that will leave the reader understanding *how* the code sums up probabilities from the target vocabulary and from pointed words in the source sentence.

Note also that the [documentation for MXNet's `scatter_nd()` function](http://mxnet.apache.org/api/python/symbol/symbol.html#mxnet.symbol.scatter_nd) has a call-out claiming the behavior is undefined if there are duplicate indices, which can definitely occur. Do you have any idea of how to address this?

**Comment by lambdaofgod** [comment]
@Zarana-Parekh do you have any examples on running architectures with copy component? Did you succeed in running configuration from pointer-generator paper? What would be required parameters?

**Comment by fhieber** [comment]
Closing this in favor of a similar approach to pointer networks (#697), which is already merged.

**Pull request closed without merging**